### DISCUSSION TITLE
Commit range filter doesn't seem to work as expected

### DISCUSSION THREAD

**Discussion by szszoke** [vercel/turborepo#8975] [Help] [answered]
### Summary

Hello,

I have an NPM workspace with a few applications and a few internal packages.

I want to write a script that invokes a command for each package that needs to be built and I want to only consider changes between the head of the branch and a specific commit hash.

Here is a sample repo: https://github.com/szszoke/turbo-commit-range-filter
There is a Devcontainer as well.

The repo contains three projects, `a`, `b` and `c`.

`a` is an internal package that is a dependency of `c`. `b` has no dependencies.

I am running `npx turbo build --filter="...[e74874a]" --dry=json | jq .packages` and I get the following output:

```bash
$ npx turbo build --filter="...[e74874a]" --dry=json | jq .packages
turbo 2.0.12

[
  "@turbo-commit-range-filter/b"
]
```

I would expect to see both `b` and `c` in the output but I can only see `b`.

I am expecting `c` as well because I made a change to `a` in `e74874a` which is a dependency of `c`.

The truth is that I started digging because in my main project (which is unfortunately private and I cannot share it) filtering by a commit hash just resulted in all packages being built, regardless if there were changes in the commits.

### Example

https://github.com/szszoke/turbo-commit-range-filter

**Comment by chris-olszewski** [comment, marked as answer]
I believe the filter you want targets the commit before the one you provided e.g. 
```
npx turbo build --filter="...[469ae9a79]" --dry=json | jq .packages
turbo 2.0.12

[
  "//",
  "@turbo-commit-range-filter/a",
  "@turbo-commit-range-filter/b",
  "@turbo-commit-range-filter/c"
]
```

`a` is returned as it is changed and `//` is returned since the root file `README.md` was changed.

SCM filters calculate changes *between* two commits. You can see what changes are considered by using `git diff` e.g.
```
[0 olszewski@chriss-mbp] /tmp/turbo-commit-range-filter $ git diff e74874a                                                   
diff --git a/README.md b/README.md
index e69de29..1999000 100644
--- a/README.md
+++ b/README.md
@@ -0,0 +1,8 @@
+```bash
+npx turbo build --filter="...[e74874a]" --dry=json | jq .packages
+turbo 2.0.12
+
+[
+  "@turbo-commit-range-filter/b"
+]
+```
\ No newline at end of file
diff --git a/packages/b/index.ts b/packages/b/index.ts
index 23b2cbd..a9e3684 100644
--- a/packages/b/index.ts
+++ b/packages/b/index.ts
@@ -1 +1 @@
-console.log("b updated");
+console.log("b updated again");
[0 olszewski@chriss-mbp] /tmp/turbo-commit-range-filter $ git diff 469ae9a79
diff --git a/README.md b/README.md
index e69de29..1999000 100644
--- a/README.md
+++ b/README.md
@@ -0,0 +1,8 @@
+```bash
+npx turbo build --filter="...[e74874a]" --dry=json | jq .packages
+turbo 2.0.12
+
+[
+  "@turbo-commit-range-filter/b"
+]
+```
\ No newline at end of file
diff --git a/packages/a/index.ts b/packages/a/index.ts
index 3f982bb..6e9da35 100644
--- a/packages/a/index.ts
+++ b/packages/a/index.ts
@@ -1,3 +1,3 @@
 export function doWork() {
-  console.log("a");
+  console.log("a updated");
 }
\ No newline at end of file
diff --git a/packages/b/index.ts b/packages/b/index.ts
index 23b2cbd..a9e3684 100644
--- a/packages/b/index.ts
+++ b/packages/b/index.ts
@@ -1 +1 @@
-console.log("b updated");
+console.log("b updated again");
```
